kolibri/utils/server.py

```diff
@@ -405,6 +406,48 @@ def EXIT(self):
             pass
 
 
+class SystemdNotifyPlugin(SimplePlugin):
+    """
+    A plugin to notify systemd of the process' state when it's starting up and
+    shutting down.
+
+    This allows systemd to wait before starting dependent processes until all
+    Kolibri process plugins have started successfully. In particular, zeroconf
+    registration can take a few seconds.
+
+    If Kolibri is not running under systemd, this plugin will do nothing.
+    """
+
+    def __init__(self, bus):
+        self.bus = bus
+        self.notify_socket_path = os.environ.get("NOTIFY_SOCKET", None)
+
+        self.bus.subscribe("RUN", self.send_ready, priority=999)
+        self.bus.subscribe("STOP", self.send_stopping, priority=1)
+        self.bus.subscribe("EXIT", self.send_stopping, priority=1)
+
+    def sd_notify(self, state):
+        """
+        Sends a state notification to systemd
+
+        See man page sd_notify(3)
+
+        :param: state: new service state
+        """
+        if not self.notify_socket_path:
+            return
+        with socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM) as s:
```

Comment:
Are there any errors that could occur here that might need to be handled?

Reply:
Generically handled all `OSError`s